_source/_posts/2022-04-05-state-of-ffi-java.md

```diff
@@ -0,0 +1,280 @@
+- Use native libraries
+- Access functions or programs on the host OS
+- Multi-precision arithmetic, Matrix multiplications
+- GPU and CPU offloading (Cuda, OpenCL, OpenGL, Vulcan, DirectX, and so on)
+- Deep learning (Tensorflow, cuDNN, Blas, and so on)
+- OpenSSL, V8, and many more
+
+## A brief history of FFI in Java
+
+Before we dwell on the current state of FFI in Java, let's look at a brief history of FFI in Java.
+
+### Java Native Interface (JNI)
+
+For a long time, the standard for FFI in Java has been Java Native Interface (JNI), and it is notorious for being slow and insecure. If you are used to other languages like Rust, Go, or Python, you will probably know how easy and intuitive it is to use FFI in them and that leaves something to be desired in Java. Even to do a small native call using JNI, you have to do a considerable amount of work, and it could still go wrong and end up being a security issue for the app.
+
+The main issues with JNI are its complexity to use and the need to write C bridge code manually. This could lead to unsafe code and pose security risks. It can also cause performance overhead in some situations.
+The performance and memory safety of the JNI code depends on the developer, and hence the mileage will vary.
+
+**Pros**
+
+- Native interface access for C/C++/Assembly
+- Fastest solution in Java
+
+**Cons**
+
+- Complicated to use and brittle
+- Not very secure and could cause memory safety issues
+- Overhead and performance loss is possible
+- Difficult to debug
+- Depends on Java developers to write safe C binding code manually
+
+### Java Native Access (JNA)
+
+The complexity of JNI gave rise to some community-driven libraries that made it simpler to do FFI in Java. [Java Native Access (JNA)](https://github.com/java-native-access/jna) is one of them. It's built on top of JNI and at least makes FFI easier to use, especially it removes the need to write any C binding code manually and reduces the chances of memory safety issues. It still has some disadvantages of being JNI-based and is slightly slower than JNI in many cases. It's also widely used and battle-tested, so definitely a better option than using JNI directly.
+
+**Pros**
+
+- Native interface access for C/C++/Assembly
+- Simpler to use compared to JNI
+- Dynamic binding. No need to write any C binding code manually
+- Widely used and mature library
+
+**Cons**
+
+- Uses reflection
+- Built on top of JNI
+- Has performance overhead and can be slower than JNI
+- Difficult to debug
+
+### Java Native Runtime (JNR)
+
+Another popular option is [Java Native Runtime (JNR)](https://github.com/jnr/jnr-ffi). Though not as widely used or mature as JNA, it's much more modern and has better performance than JNA for most use cases. However, there are some cases where JNA might perform better.
+
+**Pros**
+
+- Native interface access for C/C++/Assembly
+- Easy to use
+- Dynamic binding. No need to write any C binding code manually
```

Comment:
```suggestion
- Dynamic binding, no need to write any C binding code manually
```
to avoid a multi-sentence format, for consistency with the other bullets